### PR TITLE
Origin/ft signatory impl

### PR DESCRIPTION
**Pull request by mubarak23**

### CODE REVIEW

File: tests/test_permissionable_component.cairo [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    let (account_contract_address, _) = account_contract
+        .deploy(@acct_constructor_calldata)
+        .unwrap();
+
+    (account_contract_address, erc721_contract_address)
+}
+
+#[test]
+#[should_panic(expected: ('Account: invalid length',))]
+fn test_when_permissioned_addresses_and_permissions_not_equal() {
```
Comment (Darlington02) [inline tests/test_permissionable_component.cairo:78]: rename test to `test_should_fail_if_unequal_permissioned_addresses_and_permissions`

File: tests/test_permissionable_component.cairo [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+    let permissionable_dispatcher = IPermissionableDispatcher { contract_address };
+
+    start_cheat_caller_address(contract_address, owner);
+    permissionable_dispatcher.set_permission(permission_addresses, permissions)
+}
+
+
+#[test]
+fn test_permissionable() {
```
Comment (Darlington02) [inline tests/test_permissionable_component.cairo:101]: rename test to `test_set_permission`

File: tests/test_permissionable_component.cairo [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+    let has_permission = permissionable_dispatcher
+        .has_permission(owner, ACCOUNT2.try_into().unwrap());
+
+    assert(has_permission == true, 'Account: not permitted');
+    stop_cheat_caller_address(contract_address);
+}
+
+#[test]
+fn test_not_permitted() {
```
Comment (Darlington02) [inline tests/test_permissionable_component.cairo:130]: rename test to `test_has_permissions`

File: tests/test_permissionable_component.cairo [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    let has_permission = permissionable_dispatcher
+        .has_permission(owner, ACCOUNT4.try_into().unwrap());
+
+    assert(has_permission == false, 'Account: permitted');
+    stop_cheat_caller_address(contract_address);
+}
+
+
+#[test]
+fn test_permissionable_emit_event() {
```
Comment (Darlington02) [inline tests/test_permissionable_component.cairo:160]: rename to `test_set_permission_emits_event`

File: tests/test_permissionable_component.cairo [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+    start_cheat_caller_address(contract_address, owner);
+
+    let permissionable_dispatcher = IPermissionableDispatcher { contract_address };
+    permissionable_dispatcher.set_permission(permission_addresses, permissions);
+
+    let has_permission = permissionable_dispatcher
+        .has_permission(owner, ACCOUNT4.try_into().unwrap());
+
+    assert(has_permission == false, 'Account: permitted');
```
Comment (Darlington02) [inline tests/test_permissionable_component.cairo:154]: add checks for all 3 accounts in the array, not just ACCOUNT4

File: src/components/permissionable/permissionable.cairo [hunk 6]
```diff
@@ -1,3 +1,112 @@
...
+        pub owner: ContractAddress,
+        pub permissioned_address: ContractAddress,
+        pub has_permission: bool,
+    }
+
+    // *************************************************************************
+    //                              ERRORS
+    // *************************************************************************
+    pub mod Errors {
+        pub const UNAUTHORIZED: felt252 = 'Account: unauthorized';
```
Comment (Darlington02) [inline src/components/permissionable/permissionable.cairo:53]: remove `UNAUTHORIZED`, `NOT_OWNER`, `NOT_PERMITTED` errors, as they are not used anywhere in this component

**Review by Darlington02 — CHANGES_REQUESTED** [review]

File: tests/test_permissionable_component.cairo (hunk 5 above)
Comment (mubarak23) [reply tests/test_permissionable_component.cairo:154]: ok

File: src/components/permissionable/permissionable.cairo [hunk 7]
```diff
@@ -1,3 +1,109 @@
...
+        impl Account: AccountComponent::HasComponent<TContractState>
+    > of IPermissionable<ComponentState<TContractState>> {
+        // @notice sets permission for an account
+        // @permissioned_addresses array of addresses who's permission is to be updated
+        // @param permssions permission value <true, false>
+        fn set_permission(
+            ref self: ComponentState<TContractState>,
+            permissioned_addresses: Array<ContractAddress>,
+            permissions: Array<bool>
+        ) {
```
Comment (Darlington02) [inline src/components/permissionable/permissionable.cairo:74]: Was going over our implementation again and it just came to mind that we should enforce only the contract owner calling this function. else permissioned addresses might give permissions to random addresses on the owner's account.

So let's enforce an assert that checks that the caller is the account owner.

File: src/components/presets/account_preset.cairo [hunk 8]
```diff
@@ -90,10 +132,41 @@ pub mod AccountPreset {
...
+    #[abi(embed_v0)]
+    impl Permissionable of IPermissionable<ContractState> {
+        fn set_permission(
+            ref self: ContractState,
+            permissioned_addresses: Array<ContractAddress>,
+            permissions: Array<bool>
+        ) {
+            // validate signer
+            let caller = get_caller_address();
+            assert(self.is_valid_signer(caller), 'Account: unauthorized');
```
Comment (Darlington02) [inline src/components/presets/account_preset.cairo:160]: folllowing my earlier comment on enforcing only owner to call the `set_permission` function, this check for valid signer becomes unnecessary and redundant. Let's take it off

File: tests/test_permissionable_component.cairo (hunk 1 above)
Comment (Darlington02) [reply tests/test_permissionable_component.cairo:78]: this is still yet to be renamed

File: tests/test_permissionable_component.cairo (hunk 2 above)
Comment (Darlington02) [reply tests/test_permissionable_component.cairo:101]: this is still yet to be renamed

**Review by Darlington02 — CHANGES_REQUESTED** [review]

**Comment by mubarak23** [comment]
@Darlington02 the signature component did not implement an interface 

`ISignatory`, 

its uses  `#[generate_trait]` which the compiler  generate the corresponding trait definition, 

am trying to figure out a way to access the trait definition as dispatchers inside my test.

**Comment by Darlington02** [comment]
> @Darlington02 the signature component did not implement an interface
> 
> `ISignatory`,
> 
> its uses `#[generate_trait]` which the compiler generate the corresponding trait definition,
> 
> am trying to figure out a way to access the trait definition as dispatchers inside my test.

Just similar to how upgradeable component is tested. You test against the preset implementation. Checking that the implementation logic within the component stands. There’s a commented section of the account component for testing is_valid_signature you could utilize. Just copy and modify to include test for root_owner too.


Also since the implementation of is_valid_signer uses the permissions valid signer that contains other logic, if that works, then you can assume other implementation of is_valid_signer works. So test against those functions within preset: is_valid_signer and is_valid_signature.

**Comment by mubarak23** [comment]
> > @Darlington02 the signature component did not implement an interface
> > `ISignatory`,
> > its uses `#[generate_trait]` which the compiler generate the corresponding trait definition,
> > am trying to figure out a way to access the trait definition as dispatchers inside my test.
> 
> Just similar to how upgradeable component is tested. You test against the preset implementation. Checking that the implementation logic within the component stands. There’s a commented section of the account component for testing is_valid_signature you could utilize. Just copy and modify to include test for root_owner too.
> 
> Also since the implementation of is_valid_signer uses the permissions valid signer that contains other logic, if that works, then you can assume other implementation of is_valid_signer works. So test against those functions within preset: is_valid_signer and is_valid_signature.

am not sure if the test i wrote follow this. 

kindly review

File: src/components/presets/account_preset.cairo (hunk 8 above)
Comment (Darlington02) [reply src/components/presets/account_preset.cairo:160]: this hasn't been resolved too

**Review by Darlington02 — CHANGES_REQUESTED** [review]

File: tests/test_signatory_component.cairo [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    start_cheat_caller_address(contract_address, token_owner);
+    permissionable_dispatcher.set_permission(permission_addresses, permissions);
+    let is_valid_signer = signatory_dispatcher.is_valid_signer(ACCOUNT2.try_into().unwrap());
+
+    assert(is_valid_signer == true, 'should accept valid signer');
+    stop_cheat_caller_address(contract_address);
+}
+
+#[test]
+fn test_is_valid_signer_root_owner() {
```
Comment (Darlington02) [inline tests/test_signatory_component.cairo:141]: wrong test implementation! you are meant to test for the root owner not the token owner! get the root owner by calling the `_get_root_owner` function on the tokenbound account and testing against that instead.

Hint: you might want to expose the `_get_root_owner` by adding it as an extra function to the preset account.

File: tests/test_signatory_component.cairo [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+    start_cheat_caller_address(contract_address, token_owner);
+    let mut good_signature = array![data.r, data.s];
+    let is_valid = signatory_dispatcher.is_valid_signature(hash, good_signature.span());
+    assert(is_valid == 'VALID', 'should accept valid signature');
+    stop_cheat_caller_address(contract_address);
+}
+
+#[test]
+fn test_is_valid_signature_root_owner() {
```
Comment (Darlington02) [inline tests/test_signatory_component.cairo:178]: wrong test implementation! you are meant to test for the root owner not the token owner! get the root owner by calling the `_get_root_owner` function on the tokenbound account and testing against that instead.

Hint: you might want to expose the `_get_root_owner` by adding it as an extra function to the preset account.

**Review by Darlington02 — CHANGES_REQUESTED** [review]

**Pull request merged**